File: packages/es-markdown-core/tests/edit.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { DiagnosticCode, Document, MarkupWalker, WalkEvent } from "../dist/index.js";

function flatten(document) {
    const nodes = [];
    new MarkupWalker().walk(document, (event, node) => {
        if (event === WalkEvent.entering) nodes.push(node);
    });
    return nodes;
}

/**
 * The nodes an edit stamped: everything in `successor` whose revision is
 * newer than the whole predecessor tree. The predecessor's root revision
 * bounds its every node, because a child's revision never exceeds its
 * parent's — so a node above that bound was stamped by this edit, and a node
 * at or below it was carried.
 */
function stamped(successor, predecessorRevision) {
    return flatten(successor).filter((node) => node.revision > predecessorRevision);
}

test("edits: extending the trailing text keeps its identity and bumps its revision", () => {
    const first = Document("# Title\n\nHello");
    const firstHeading = first.content[0];
    const firstParagraph = first.content[1];
    const firstText = firstParagraph.content[0];

    const second = first.edit("# Title\n\nHello world");
    const secondHeading = second.content[0];
    const secondParagraph = second.content[1];
    const secondText = secondParagraph.content[0];

    assert.equal(secondText.literal, "Hello world");
    assert.equal(secondParagraph.id, firstParagraph.id);
    assert.equal(secondText.id, firstText.id);
    assert.ok(secondText.revision > firstText.revision);
    // The heading settled before the edit: equal to its predecessor down to
    // the last field. It is not the SAME object — every node is decoded
    // every time — and nothing should need it to be: what a reactive
    // consumer compares is the (id, revision) pair.
    assert.equal(secondHeading.id, firstHeading.id);
    assert.equal(secondHeading.revision, firstHeading.revision);
    assert.deepEqual(secondHeading, firstHeading);

    // The edit stamped the text and everything above it — and only that.
    // The stamp is what says "changed"; the heading keeps its old revision.
    const touched = new Set(stamped(second, first.revision).map((node) => node.id));
    assert.equal(touched.has(secondHeading.id), false);
    assert.equal(touched.has(secondText.id), true);
    // Nothing was removed: every predecessor identity is still present.
    for (const node of flatten(first)) {
        assert.notEqual(second.node(node.id), null);
    }
    second.close();
});

test("edits: a clean-boundary insert at the top leaves downstream identity intact", () => {
    const before = Document("First\n\nSecond\n\nThird\n");
    const downstreamBefore = before.content.map((node) => [node.id, node.revision]);
    const thirdBefore = before.content[2];

    const after = before.edit("# New\n\nFirst\n\nSecond\n\nThird\n");

    assert.equal(after.content.length, 4);
    const inserted = after.content[0];
    after.content.slice(1).forEach((node, index) => {
        assert.equal(node.id, downstreamBefore[index][0]);
        assert.equal(node.revision, downstreamBefore[index][1]);
    });

    // A created node is a fresh identity carrying the new document revision.
    assert.ok(downstreamBefore.every(([id]) => id !== inserted.id));
    assert.equal(before.node(inserted.id), null);
    assert.equal(inserted.revision, after.revision);
    assert.ok(inserted.revision > before.revision);

    // Downstream nodes shifted by two lines: same identity, same revision,
    // new extent. Each value states where it is, so the predecessor's value
    // keeps reporting the predecessor's position forever — and the pair a
    // reactive consumer compares is unmoved by the shift.
    const third = after.content[3];
    assert.equal(third.scope.start.line, 7);
    assert.equal(thirdBefore.scope.start.line, 5);
    assert.equal(third.id, thirdBefore.id);
    assert.equal(third.revision, thirdBefore.revision);
    const reference = Document("# New\n\nFirst\n\nSecond\n\nThird\n");
    assert.equal(after.dump(), reference.dump());
    reference.close();
    after.close();
});

test("edits: a kind change retires the old identity and mints a new one", () => {
    const before = Document("text\n");
    const paragraph = before.content[0];

    const after = before.edit("# text\n");
    const heading = after.content[0];
    assert.notEqual(heading.id.rawValue, paragraph.id.rawValue);
    // The minted node carries the new document revision; the retired one is
    // simply gone from the tree.
    assert.equal(heading.revision, after.revision);
    assert.equal(after.node(paragraph.id), null);
    after.close();
});

test("edits: equality is series-salted identity plus revision", () => {
    const source = "Same *content* twice.\n";
    const first = Document(source);
    const second = Document(source);
    // Identical content from different parses never shares identity.
    assert.notEqual(first.id.series, second.id.series);
    assert.notEqual(first.content[0], second.content[0]);
    // An id from another series is not this document's to answer.
    assert.equal(first.node(second.content[0].id), null);
    assert.equal(first.node(first.content[0].id), first.content[0]);
    // The root answers for itself, like every node under it.
    assert.equal(first.node(first.id), first);
    first.close();
    second.close();
});

test("edits: a blank-line-only edit changes no node yet shifts scopes", () => {
    const before = Document("Alpha\n\n\n\nOmega\n");
    const omegaBefore = before.content[1];
    assert.equal(omegaBefore.scope.start.line, 5);
    const pairsBefore = flatten(before).map((node) => [node.id, node.revision]);

    // Delete two of the blank lines: no node's content changes, so every
    // node — the root included — keeps its exact identity and revision.
    // Position is not content, and revisions do not report it.
    const after = before.edit("Alpha\n\nOmega\n");
    assert.deepEqual(
        flatten(after).map((node) => [node.id, node.revision]),
        pairsBefore
    );
    assert.equal(after.content[1].id, omegaBefore.id);
    assert.equal(after.content[1].revision, omegaBefore.revision);
    assert.equal(after.content[1].scope.start.line, 3);
    const reference = Document("Alpha\n\nOmega\n");
    assert.equal(after.dump(), reference.dump());
    reference.close();
    after.close();
});

test("edits: a deep rebuild stamps the changed spine and nothing else", () => {
    const depth = 512;
    const stable = "Stable\n\n";
    const prefix = "> ".repeat(depth);
    const before = Document(stable + prefix + "alpha\n");
    const stableBefore = before.content[0];

    const after = before.edit(stable + prefix + "bravo\n");

    // The settled paragraph is untouched, so it survives byte for byte.
    assert.deepEqual(after.content[0], stableBefore);

    // One text changed, and a revision covers a node's whole subtree, so the
    // stamp reaches every ancestor: the innermost text, the chain of quotes,
    // and the root all carry the new document revision, while the settled
    // paragraph's subtree keeps its old one.
    let innermost = after.content[1];
    while (innermost.kind === "blockQuote") innermost = innermost.content[0];
    const text = innermost.content[0];
    assert.equal(text.literal, "bravo");
    assert.equal(text.revision, after.revision);
    const touched = stamped(after, before.revision);
    assert.ok(touched.length >= depth);
    const touchedIds = new Set(touched.map((node) => node.id));
    assert.equal(touchedIds.has(after.content[0].id), false);
    assert.equal(touchedIds.has(after.id), true);
    after.close();
});

test("edits: a superseded document keeps answering from its own values", () => {
    const first = Document("One\n\nTwo\n");
    const two = first.content[1];
    assert.equal(two.scope.start.line, 3);

    // The edit supersedes its receiver; here even the successor is closed at
    // once. The predecessor's values, scopes, diagnostics, and dump were all
    // extracted at parse time and owe nobody anything — decoded values live
    // forever.
    first.edit("Zero\n\nOne\n\nTwo\n").close();
    assert.equal(two.scope.start.line, 3);
    assert.ok(first.dump().includes("Paragraph"));
    assert.equal(flatten(first).length, 5);
    first.close();
});

test("edits: a mutation supersedes its receiver, so history is linear", () => {
    // A document is the live head of a chain. A mutation advances the chain
    // and supersedes its receiver: mutating the old head again is a
    // deterministic error, never a second line of descent.
    const base = Document("One\n");
    const head = base.edit("Two\n");
    assert.equal(head.content[0].content[0].literal, "Two");
    assert.throws(() => base.edit("Three\n"), /superseded/);
    assert.throws(() => base.append("Three\n"), /superseded/);
    // The rejected calls changed nothing: the real head still mutates, and
    // the superseded document still answers from its decoded values.
    assert.equal(base.content[0].content[0].literal, "One");
    const next = head.edit("Three\n");
    assert.equal(next.content[0].content[0].literal, "Three");
    assert.ok(next.revision > head.revision);
    base.close();
    head.close();
    next.close();

    const closed = Document("One\n");
    closed.close();
    // Closing twice is a no-op; editing after it is not.
    closed.close();
    assert.throws(() => closed.edit("Three\n"), /released/);
    // What was already extracted stays answerable either way.
    assert.equal(closed.content[0].scope.start.line, 1);
});

test("edits: diagnostics travel with the document that raised them", () => {
    // The one thing an editor underlines: a directive's `{...}` did not
    // parse, so the braces stayed literal text. Invisible in the tree — the
    // node simply has no attributes — which is why it is reported.
    const before = Document(":::note{= bad}\nbody\n:::\n");
    assert.deepEqual(
        before.diagnostics.map((diagnostic) => diagnostic.code),
        [DiagnosticCode.directiveAttributes]
    );
    assert.deepEqual(before.diagnostics[0].scope, {
        start: { line: 1, column: 8 },
        end: { line: 1, column: 14 }
    });

    const after = before.edit(":::note{a=1}\nbody\n:::\n");
    assert.equal(after.diagnostics.length, 0);
    // The predecessor keeps reporting its own: diagnostics are values it
    // copied out at parse time, like every other part of it.
    assert.equal(before.diagnostics.length, 1);
    after.close();
});

test("edits: options are fixed for a series and reported by every revision", () => {
    const before = Document("| a |\n| --- |\n| b |\n", { tables: false });
    assert.equal(before.options.tables, false);
    assert.equal(before.content[0].kind, "paragraph");
    const after = before.edit("| a |\n| --- |\n| c |\n");
    assert.equal(after.options.tables, false);
    assert.equal(after.content[0].kind, "paragraph");
    after.close();
});

test("edits: irregular render ticks over a multi-turn conversation", () => {
    // The shape of a real LLM consumer: every socket message APPENDS to the
    // document — the hot path — and an irregular render tick only verifies
    // what a renderer would draw. Three assistant turns extend one document;
    // blocks settled at a turn boundary must stay frozen while later turns
    // stream.
    const turns = [
        "# Streaming\n\nThe *quick* parser holds **steady** under bursts, " +
            "and the heading keeps its identity from the first render on.\n\n" +
            "Update traffic stays proportional to what changed, so a renderer " +
            "reconciles by id instead of walking the whole tree.\n\n" +
            "> Snapshots are values: whatever a tick captured stays valid " +
            "while the socket races ahead.",
        "\n\n- append per message\n- edit per tick\n- settled blocks stay frozen" +
            "\n- identical items stress identity\n- identical items stress identity" +
            "\n\n```swift\nlet constant = 1\nlet mirror = [Int: String]()\n" +
            "for index in 0..<3 {\n    print(index, constant)\n}\n```\n\n" +
            "Fenced code arrives line by line and only closes at the final tick.",
        "\n\nA table lands late in the conversation:\n\n" +
            "| stage | commits | messages |\n| - | - | - |\n| one | 3 | 9 |\n" +
            "| two | 5 | 14 |\n| three | 8 | 21 |\n\n" +
            "Tail with a footnote[^n] whose definition arrives last.\n\n" +
            "[^n]: Resolved at the end, after every reference already rendered."
    ];
    // One fixed generator drives batch sizes and tick timing, so the burst
    // shapes are irregular but reproducible — and identical in the Swift and
    // Kotlin mirrors of this test.
    let state = 0x9e3779b97f4a7c15n;
    const draw = (bound) => {
        state = BigInt.asUintN(64, state * 6364136223846793005n + 1442695040888963407n);
        return Number((state >> 33n) % bound);
    };

    let document = Document("");
    let streamed = "";
    let frozen = [];
    let messages = 0;
    let ticks = 0;
    let touched = 0;

    const tick = () => {
        ticks += 1;
        const reference = Document(streamed);
        assert.equal(document.dump(), reference.dump());
        reference.close();
        for (const [index, id, revision] of frozen) {
            assert.equal(document.content[index].id, id);
            assert.equal(document.content[index].revision, revision);
        }
    };

    for (const turn of turns) {
        let offset = 0;
        while (offset < turn.length) {
            // Mostly a 20-30 token batch (80-150 characters), with
            // occasional tiny flushes of a few words. Cuts land at raw
            // character offsets — mid-word, mid-marker, even between the two
            // newlines of a block boundary — because that is the steady
            // state of LLM output.
            const width = draw(10n) < 2 ? 2 + draw(18n) : 80 + draw(71n);
            const chunk = turn.slice(offset, offset + width);
            streamed += chunk;
            offset += width;
            messages += 1;
            const previous = document;
            document = document.append(chunk);
            // What this append stamped, measured off the new tree itself:
            // the nodes carrying a fresher revision than the whole
            // predecessor tree — the same proxy the Swift and Kotlin mirrors
            // of this test count.
            touched += stamped(document, previous.revision).length;
            previous.close();
            if (draw(4n) === 0) tick();
        }
        // The turn boundary always renders; everything but the still-hot
        // last block is now settled.
        tick();
        frozen = document.content.slice(0, -1).map((node, index) => [index, node.id, node.revision]);
    }
    assert.ok(messages > 9);
    assert.ok(ticks < messages);
    const reference = Document(turns.join(""));
    assert.equal(document.dump(), reference.dump());
    reference.close();

    // Near-O(n) pipeline: total stamped-node traffic stays within one node
    // per final node plus bounded frontier churn per append. A full rebuild
    // per message would be on the order of messages * nodes.
    assert.ok(touched < flatten(document).length + 24 * messages);
    document.close();
});

File: packages/kotlin-markdown-core/src/commonTest/kotlin/com/nouprax/markdown/core/EditTest.kt
package com.nouprax.markdown.core

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertIs
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class EditTest {
    @Test
    fun extendingTheTrailingTextKeepsItsIdentityAndBumpsItsRevision() {
        val first = Document("# Title\n\nHello")
        val firstHeading = assertIs<Heading>(first.content[0])
        val firstParagraph = assertIs<Paragraph>(first.content[1])
        val firstText = assertIs<Text>(firstParagraph.content[0])

        first.edit("# Title\n\nHello world").use { second ->
            val secondHeading = assertIs<Heading>(second.content[0])
            val secondParagraph = assertIs<Paragraph>(second.content[1])
            val secondText = assertIs<Text>(secondParagraph.content[0])

            assertEquals("Hello world", secondText.literal)
            assertEquals(firstParagraph.id, secondParagraph.id)
            assertEquals(firstText.id, secondText.id)
            assertTrue(secondText.revision > firstText.revision)
            assertEquals<Markup>(firstHeading, secondHeading)

            // The heading settled before the edit and keeps its exact
            // revision; the changed text and the paragraph above it carry the
            // new document revision — which is how "what changed" is asked of
            // the tree itself.
            assertEquals(firstHeading.revision, secondHeading.revision)
            assertEquals(second.revision, secondText.revision)
            assertEquals(second.revision, secondParagraph.revision)
        }
    }

    @Test
    fun aCleanBoundaryInsertAtTheTopLeavesDownstreamIdentityIntact() {
        val before = Document("First\n\nSecond\n\nThird\n")
        val downstreamBefore = before.content.map { it.id to it.revision }
        val thirdBefore = assertIs<Paragraph>(before.content[2])

        before.edit("# New\n\nFirst\n\nSecond\n\nThird\n").use { after ->
            assertEquals(4, after.content.size)
            val inserted = assertIs<Heading>(after.content[0])
            after.content.drop(1).forEachIndexed { index, node ->
                assertEquals(downstreamBefore[index].first, node.id)
                assertEquals(downstreamBefore[index].second, node.revision)
            }

            // A created node carries the new document revision, exactly like
            // everything else this edit minted or changed.
            assertEquals(after.revision, inserted.revision)

            // Downstream nodes shifted by two lines: equal values, new scopes.
            val third = assertIs<Paragraph>(after.content[3])
            assertEquals(7, third.scope.start.line)
            // The predecessor's value keeps the extent it was parsed with, and
            // the two are still EQUAL: identity survived the edit and position
            // is not part of equality.
            assertEquals(5, thirdBefore.scope.start.line)
            assertEquals<Markup>(thirdBefore, third)
            assertEquals(Document("# New\n\nFirst\n\nSecond\n\nThird\n").dump(), after.dump())
        }
    }

    @Test
    fun aKindChangeRetiresTheOldIdentityAndMintsANewOne() {
        val before = Document("text\n")
        val paragraph = assertIs<Paragraph>(before.content[0])

        before.edit("# text\n").use { after ->
            val heading = assertIs<Heading>(after.content[0])
            assertNotEquals(paragraph.id, heading.id)

            // The minted identity carries the new document revision, and
            // retirement is a presence question: the old id is simply no
            // longer this series' to answer.
            assertEquals(after.revision, heading.revision)
            assertNull(after.node(paragraph.id))
            // The predecessor still answers for it — retirement is the
            // successor's fact, not a rewrite of history.
            assertEquals(paragraph.id, before.node(paragraph.id)?.id)
        }
    }

    @Test
    fun equalityIsSeriesSaltedIdentityPlusRevision() {
        Document("Same *content* twice.\n").use { first ->
            Document("Same *content* twice.\n").use { second ->
                // Identical content from different parses never compares equal.
                assertNotEquals<Markup>(first, second)
                assertNotEquals<Markup>(first.content[0], second.content[0])
                // Within one document, identity is value equality.
                assertEquals(first.content[0], first.content[0])
                assertNotEquals(first.id.series, second.id.series)
                // An id from another series is not this document's to answer.
                assertNull(first.node(second.content[0].id))
            }
        }
    }

    @Test
    fun aBlankLineOnlyEditChangesNoRevisionYetShiftsScopes() {
        val before = Document("Alpha\n\n\n\nOmega\n")
        val omegaBefore = assertIs<Paragraph>(before.content[1])
        assertEquals(5, omegaBefore.scope.start.line)

        // Delete two of the blank lines: no node's content changes.
        before.edit("Alpha\n\nOmega\n").use { after ->
            val omegaAfter = assertIs<Paragraph>(after.content[1])
            // Nothing changed anywhere, so even the root kept its exact
            // revision — the tree's way of reporting an empty change set.
            assertEquals(before.revision, after.revision)
            assertEquals(before.id, after.id)
            // Equal, and at a different place: two nodes differing only in
            // where they sit are equal, because position is not content.
            assertEquals<Markup>(omegaBefore, omegaAfter)
            assertEquals(3, omegaAfter.scope.start.line)
            assertEquals(5, omegaBefore.scope.start.line)
            assertEquals(Document("Alpha\n\nOmega\n").dump(), after.dump())
        }
    }

    @Test
    fun aDeepRebuildStampsEveryAncestorOfTheOneChangedText() {
        val depth = 512
        val stable = "Stable\n\n"
        val prefix = "> ".repeat(depth)
        val before = Document(stable + prefix + "alpha\n")
        val stableBefore = assertIs<Paragraph>(before.content.first())

        before.edit(stable + prefix + "bravo\n").use { after ->
            assertEquals(stableBefore.id, after.content.first().id)
            assertEquals(stableBefore.revision, after.content.first().revision)
            assertEquals(Document(stable + prefix + "bravo\n").dump(), after.dump())

            // One text changed, and a revision covers the whole subtree below
            // its node: the changed text, the paragraph and every quote above
            // it, and the document itself carry the new document revision,
            // while the settled paragraph keeps its own — which is what lets
            // a renderer prune at the first node whose revision did not move.
            var innermost: Markup = after.content.last()
            while (innermost is BlockQuote) {
                assertEquals(after.revision, innermost.revision)
                innermost = innermost.content.first()
            }
            val text = assertIs<Text>(assertIs<Paragraph>(innermost).content.first())
            assertEquals("bravo", text.literal)
            assertEquals(after.revision, text.revision)
            // The chain, the paragraph it ends in, its text, and the document:
            // exactly depth + 3 nodes carry the new revision, and nothing else
            // does.
            var stamped = 0
            MarkupWalker.walk(after) { event, node, _ ->
                if (event == WalkEvent.ENTERING && node.revision == after.revision) {
                    stamped += 1
                }
            }
            assertEquals(depth + 3, stamped)
        }
    }

    @Test
    fun aSupersededDocumentKeepsAnsweringFromItsOwnTables() {
        val first = Document("One\n\nTwo\n")
        val two = assertIs<Paragraph>(first.content[1])
        assertEquals(3, two.scope.start.line)

        // The edit supersedes the receiver — mutation is now the successor's
        // alone — but supersession severs mutation, not the values. Content,
        // scopes, diagnostics, and dump were all extracted at parse time and
        // owe the native parse nothing, so the superseded document keeps
        // answering from its own tables.
        first.edit("Zero\n\nOne\n\nTwo\n").close()
        assertEquals(3, two.scope.start.line)
        assertTrue(first.dump().contains("Paragraph"))
        val recording = RecordingVisitor()
        MarkupWalker.walk(first, recording)
        assertEquals(5, recording.visited.size)
        first.close()
    }

    @Test
    fun aMutationSupersedesItsReceiverAndHistoryStaysLinear() {
        // A mutation advances the chain, old handles die, decoded values
        // live forever. There is no forking: mutating a superseded document
        // is a deterministic refusal with the engine's own wording, and the
        // refusal disturbs nothing.
        val base = Document("One\n")
        val head = base.edit("Two\n")
        for (refused in listOf<() -> Document>({ base.edit("Three\n") }, { base.append("!") })) {
            val refusal = assertFailsWith<IllegalStateException>(block = refused)
            assertEquals("the document has been superseded: mutate the successor", refusal.message)
        }
        // The superseded document still answers from its decoded values...
        assertEquals("One", assertIs<Text>(assertIs<Paragraph>(base.content[0]).content[0]).literal)
        // ...and the head is untouched by the refusals: it is still the live
        // head, and both mutations remain its to make.
        assertEquals("Two", assertIs<Text>(assertIs<Paragraph>(head.content[0]).content[0]).literal)
        val appended = head.append("Three\n")
        assertEquals(Document("Two\nThree\n").use { it.dump() }, appended.dump())
        assertTrue(appended.revision > head.revision)
        appended.close()
        head.close()
        base.close()

        val closed = Document("One\n")
        closed.close()
        // Closing twice is a no-op; mutating after it is not, and the closed
        // refusal is its own message — closing is the caller's act, not a
        // mutation's.
        closed.close()
        val refusal = assertFailsWith<IllegalStateException> { closed.edit("Three\n") }
        assertEquals("the document is closed", refusal.message)
        // What was already extracted stays answerable either way.
        assertEquals(
            1,
            closed.content[0]
                .scope.start.line,
        )
    }

    @Test
    fun documentsAreValuesAndIdsAreStableMapKeys() {
        Document("").use { empty ->
            assertTrue(empty.content.isEmpty())
            empty.edit("Alpha\n").use { document ->
                val paragraph = assertIs<Paragraph>(document.content[0])
                val byId = hashMapOf(paragraph.id to "paragraph")
                assertEquals(paragraph.id, document.node(paragraph.id)?.id)
                // The root answers for itself, which is what a consumer
                // reconciling by id needs when the top-level list changes.
                assertEquals(document.id, document.node(document.id)?.id)
                assertEquals("paragraph", byId[paragraph.id])
                assertNull(document.node(MarkupID(document.series + 1UL, paragraph.id.rawValue)))
            }
        }
    }

    @Test
    fun diagnosticsTravelWithTheDocumentThatRaisedThem() {
        // The one thing an editor underlines: a directive's `{...}` did not
        // parse, so the braces stayed literal text. Invisible in the tree —
        // the node simply has no attributes — which is why it is reported.
        val before = Document(":::note{= bad}\nbody\n:::\n")
        assertEquals(listOf(DiagnosticCode.DIRECTIVE_ATTRIBUTES), before.diagnostics.map { it.code })
        assertEquals(Scope(Position(1, 8), Position(1, 14)), before.diagnostics.single().scope)

        before.edit(":::note{a=1}\nbody\n:::\n").use { after ->
            assertTrue(after.diagnostics.isEmpty())
        }
        // The predecessor keeps reporting its own: diagnostics are values it
        // copied out at parse time, like every other part of it.
        assertEquals(1, before.diagnostics.size)
    }

    @Test
    fun streamingWithIrregularTicksOverAMultiTurnConversation() {
        // The shape of a real LLM consumer: every socket message rides the
        // real append — the streaming mutation — and an irregular render
        // tick reads what the appends produced. Three assistant turns extend
        // one document; blocks settled at a turn boundary must stay frozen
        // while later turns stream. The corpus text and the split generator
        // are the shared burst shapes, byte-identical in the Swift and ES
        // mirrors of this test.
        val splits = StreamingCorpus.Splits()

        var document = Document("")
        var streamed = ""
        var frozen = emptyList<Triple<Int, MarkupID, ULong>>()
        var messages = 0
        var ticks = 0
        var touched = 0

        fun tick() {
            ticks += 1
            // A tick no longer parses anything: the appends already did, and
            // the tick verifies their sum against a one-shot parse of every
            // byte so far.
            Document(streamed).use { reference ->
                assertEquals(reference.dump(), document.dump())
            }
            for ((index, id, revision) in frozen) {
                val node = document.content[index]
                assertEquals(id, node.id)
                assertEquals(revision, node.revision)
            }
        }

        for (turn in StreamingCorpus.turns) {
            var offset = 0
            while (offset < turn.length) {
                val width = splits.width()
                val message = turn.substring(offset, minOf(offset + width, turn.length))
                offset += message.length
                streamed += message
                messages += 1

                // Each message advances the chain; the superseded predecessor
                // is closed behind it — an O(1) release.
                val predecessorRevision = document.revision
                val superseded = document
                document = document.append(message)
                superseded.close()

                // The old delta row count, re-expressed on the tree: the
                // nodes whose revision equals the new document revision are
                // exactly the ones this append changed or minted — and a root
                // that kept its revision means the append changed nothing.
                if (document.revision != predecessorRevision) {
                    val stamped = document.revision
                    MarkupWalker.walk(document) { event, node, _ ->
                        if (event == WalkEvent.ENTERING && node.revision == stamped) {
                            touched += 1
                        }
                    }
                }
                if (splits.draw(4L) == 0L) {
                    tick()
                }
            }
            // The turn boundary always renders; everything but the still-hot
            // last block is now settled.
            tick()
            frozen =
                document.content.dropLast(1).mapIndexed { index, node ->
                    Triple(index, node.id, node.revision)
                }
        }
        assertTrue(messages > 9)
        assertTrue(ticks < messages)
        assertEquals(
            Document(StreamingCorpus.turns.joinToString(separator = "")).use { it.dump() },
            document.dump(),
        )

        // Near-O(n) pipeline: the total count of nodes stamped with a new
        // document revision stays within one per final node plus bounded
        // frontier churn per append. A full re-decode per append would be on
        // the order of messages * nodes — which is exactly the term the
        // pruned decode deletes.
        var nodes = 0
        MarkupWalker.walk(document) { event, _, _ ->
            if (event == WalkEvent.ENTERING) {
                nodes += 1
            }
        }
        assertTrue(touched < nodes + 16 * messages)
        document.close()
    }
}
